Sources/Common/Communication/EventBus.swift

```diff
@@ -0,0 +1,106 @@
+    ///   - action: The action to execute when the event is observed.
+    func addObserver(_ eventType: String, action: @escaping (AnyEventRepresentable) -> Void) async
+    /// Removes all registered observers from the EventBus.
+    func removeAllObservers() async
+    /// Removes all observers for a specific event type.
+    ///
+    /// - Parameter eventType: The event type for which to remove observers.
+    func removeObserver(for eventType: String) async
+}
+
+/// A shared implementation of `EventBus` using an actor model for thread-safe operations.
+/// This actor manages the distribution of events to registered observers and uses
+/// `NotificationCenter` for event delivery. It ensures that event handling is thread-safe
+/// and observers are managed efficiently.
+// sourcery: InjectRegisterShared = "EventBus"
+// sourcery: InjectSingleton
+actor SharedEventBus: EventBus {
+    private var notificationCenter: NotificationCenter = .default
+    private var observers: [String: [NSObjectProtocol]] = [:]
+
+    deinit {
+        // Clean up by removing all observers when the EventBus is deinitialized.
+        // Capture a weak reference to self to avoid retain cycles
+        Task { [weak self] in
+            await self?.removeAllObservers()
+        }
+    }
+
+    init() {
+        DIGraphShared.shared.logger.debug("SharedEventBus initialized")
+    }
+
+    /// Posts an event to all registered observers of its type.
+    ///
+    /// - Parameter event: The event to be posted.
+    /// - Returns: True if the event has been posted to any observers, false otherwise.
+    @discardableResult
+    func post(_ event: AnyEventRepresentable) async -> Bool {
+        let key = event.key
+        if let observerList = observers[key], !observerList.isEmpty {
+            notificationCenter.post(name: NSNotification.Name(key), object: event)
+            return true
+        }
+        return false
+    }
+
+    /// Registers an observer for a specific event type.
+    ///
+    /// - Parameters:
+    ///   - eventType: The type of the event to observe.
+    ///   - action: The action to be executed when the event is received.
+    func addObserver(_ eventType: String, action: @escaping (AnyEventRepresentable) -> Void) {
+        let observer = notificationCenter.addObserver(forName: NSNotification.Name(eventType), object: nil, queue: nil) { notification in
+            if let event = notification.object as? AnyEventRepresentable {
+                action(event)
+            }
+        }
+        // Store the observer reference for later management.
+        if observers[eventType] != nil {
+            observers[eventType]?.append(observer)
+        } else {
+            observers[eventType] = [observer]
+        }
+    }
+
+    /// Removes all observers from the EventBus.
+    ///
+    /// This function is used for cleanup or resetting the event handling system.
+    func removeAllObservers() async {
```

Comment:
no need to specify `async` explicitly in the actor implementation. Actor methods are all async unless you mark them as `nonisolated`

Reply:
Indeed it's not needed in the implementation, but it's needed in the protocol since its a `nonisolated protocol requirement`.

If an actor now implements a protocol, automatically added method stubs will all include `async` keyword, you will need to remove them intentionally, hence left these keywords there for symmetry.

Reply:
it is non-ideal that auto-complete adds them for an actor conforming to the protocol. Including them have a semantic impact which shouldn't be added unless needed.

By default, all actor methods are `async` from the outside world. So if the protocol define `async`, they will be async without adding `async` in their actor counterpart.
However, internally it is different story. By default actor methods can call each other without `await`, which means your execution flow goes line by line as you would expect. However, if one method marked explicitly with `async` then even actor methods will have to `await` for it even if it is not needed. And it is cascading effect.

The only exception, is if the actor is calling an external api that's marked as async, then the actor calling function needs to be async in order to be able to await for the external api call

Reply:
> However, if one method marked explicitly with async then even actor methods will have to await for it even if it is not needed. And it is cascading effect.

This is a good point you raised, The actor ensures thread safety for both synchronous and asynchronous methods. The key difference is in how they are called and whether they can perform suspension points (await) and adding async would enforce us to wait for them.